dask/tests/test_multiprocessing.py

```diff
@@ -194,6 +194,38 @@ def f():
     assert len(set(results)) == N
 
 
+class global_:
+    value = 0
+
+
+def proc_init():
+    global_.value = 1
+
+
+def test_process_initializer():
+    @delayed(pure=False)
+    def f():
+        return global_.value
+
+    global_.value = 1
+    N = 10
+
```

Comment:
As this code is now, there is no problem. ***But*** ...

an unexpected error (related to pickling) occurs when I move the definition of `proc_init()` into `test_process_initializer()`.

<details>
<summary>Click to see errors obtained in such a scenario:</summary>

    ============================= test session starts =============================
    platform win32 -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0 -- C:\Users\runner\anaconda3\envs\daskenv\python.exe
    cachedir: .pytest_cache
    rootdir: C:\Users\runner\git\dask, configfile: setup.cfg
    plugins: anyio-3.5.0
    collecting ... collected 21 items

    test_multiprocessing.py::test_process_initializer Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\runner\anaconda3\envs\daskenv\lib\multiprocessing\spawn.py", line 116, in spawn_main
        exitcode = _main(fd, parent_sentinel)
      File "C:\Users\runner\anaconda3\envs\daskenv\lib\multiprocessing\spawn.py", line 126, in _main
        self = reduction.pickle.load(from_parent)
    EOFError: Ran out of input
    FAILED

    ================================== FAILURES ===================================
    __________________________ test_process_initializer ___________________________

        def test_process_initializer():
            @delayed(pure=False)
            def f():
                return global_.value
        
            def proc_init():
                global_.value = 1
        
            global_.value = 1
            N = 10
        
            with dask.config.set(scheduler="threading"):
                (results,) = compute([f() for _ in range(N)])
            expected_results = [1] * N
            assert_eq(results, expected_results)  # processes saw global_.value = 1
        
            with dask.config.set(scheduler="processes"):
                (results,) = compute([f() for _ in range(N)])
            expected_results = [0] * N
            assert_eq(results, expected_results)  # processes did not see global_.value = 1
        
            with dask.config.set(scheduler="processes", initializer=proc_init):
    >           (results,) = compute([f() for _ in range(N)])

    N          = 10
    expected_results = [0, 0, 0, 0, 0, 0, ...]
    f          = Delayed('f-24eb6044-2a4a-4d43-81a8-ef27d41b5e2d')
    proc_init  = <function test_process_initializer.<locals>.proc_init at 0x000001C9AE880F70>
    results    = [0, 0, 0, 0, 0, 0, ...]

    test_multiprocessing.py:228: 
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    ..\base.py:599: in compute
        results = schedule(dsk, keys, **kwargs)
            args       = ([Delayed('f-f1b064dd-9685-45d8-aadb-c851fb5529dd'), Delayed('f-08f5e46d-c78e-4229-a07d-2246bc072fda'), Delayed('f-9c0...99f34cc'), Delayed('f-bc67018d-f89c-45a1-90b6-856c9662a957'), Delayed('f-39aa0784-78a0-447e-9b91-cc3db878996d'), ...],)
            collections = [Delayed('f-f1b064dd-9685-45d8-aadb-c851fb5529dd'), Delayed('f-08f5e46d-c78e-4229-a07d-2246bc072fda'), Delayed('f-9c0b...3999f34cc'), Delayed('f-bc67018d-f89c-45a1-90b6-856c9662a957'), Delayed('f-39aa0784-78a0-447e-9b91-cc3db878996d'), ...]
            dsk        = HighLevelGraph with 10 layers.
    <dask.highlevelgraph.HighLevelGraph object at 0x1c9ae8616c0>
     0. f-c733873f-64c6-498c-a...1231-9fc3-4bf5-a042-fd04670f73fe
     8. f-08f5e46d-c78e-4229-a07d-2246bc072fda
     9. f-f1b064dd-9685-45d8-aadb-c851fb5529dd

            get        = None
            keys       = [['f-f1b064dd-9685-45d8-aadb-c851fb5529dd'], ['f-08f5e46d-c78e-4229-a07d-2246bc072fda'], ['f-9c0b1231-9fc3-4bf5-a042-f...-43be-afde-6593999f34cc'], ['f-bc67018d-f89c-45a1-90b6-856c9662a957'], ['f-39aa0784-78a0-447e-9b91-cc3db878996d'], ...]
            kwargs     = {}
            optimize_graph = True
            postcomputes = [(<function single_key at 0x000001C9AEAC9510>, ()), (<function single_key at 0x000001C9AEAC9510>, ()), (<function sing...C9510>, ()), (<function single_key at 0x000001C9AEAC9510>, ()), (<function single_key at 0x000001C9AEAC9510>, ()), ...]
            repack     = <function unpack_collections.<locals>.repack at 0x000001C9AE881870>
            schedule   = <function get at 0x000001C9AE975F30>
            scheduler  = None
            traverse   = True
            x          = Delayed('f-c733873f-64c6-498c-a396-1bd67a3a7c7a')
    ..\multiprocessing.py:232: in get
        result = get_async(
            chunksize  = 6
            cleanup    = True
            context    = <module 'multiprocessing' from 'C:\\Users\\runner\\anaconda3\\envs\\daskenv\\lib\\multiprocessing\\__init__.py'>
            dependencies = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': set(), 'f-39aa0784-78a0-447e-9b91-cc3db878996d': set(), 'f-3c4c16a1-469f-4008-8591-034a0c7b6f24': set(), 'f-4286a9dd-d761-4954-8bcb-02a13404601e': set(), ...}
            dsk        = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), 'f...'f-4286a9dd-d761-4954-8bcb-02a13404601e': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), ...}
            dsk2       = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), 'f...'f-4286a9dd-d761-4954-8bcb-02a13404601e': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), ...}
            dsk3       = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), 'f...'f-4286a9dd-d761-4954-8bcb-02a13404601e': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), ...}
            dumps      = functools.partial(<function dumps at 0x000001C9AE96EE60>, protocol=5)
            func_dumps = None
            func_loads = None
            initializer = <function test_process_initializer.<locals>.proc_init at 0x000001C9AE880F70>
            keys       = [['f-f1b064dd-9685-45d8-aadb-c851fb5529dd'], ['f-08f5e46d-c78e-4229-a07d-2246bc072fda'], ['f-9c0b1231-9fc3-4bf5-a042-f...-43be-afde-6593999f34cc'], ['f-bc67018d-f89c-45a1-90b6-856c9662a957'], ['f-39aa0784-78a0-447e-9b91-cc3db878996d'], ...]
            kwargs     = {}
            loads      = <built-in function loads>
            num_workers = 8
            optimize_graph = True
            pool       = <concurrent.futures.process.ProcessPoolExecutor object at 0x000001C9AE861D80>
    ..\local.py:496: in get_async
        fire_tasks(chunksize)
            _          = ()
            cache      = None
            callbacks  = ()
            chunksize  = 6
            dsk        = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), 'f...'f-4286a9dd-d761-4954-8bcb-02a13404601e': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), ...}
            dumps      = functools.partial(<function dumps at 0x000001C9AE96EE60>, protocol=5)
            fire_tasks = <function get_async.<locals>.fire_tasks at 0x000001C9AE881A20>
            get_id     = <function _process_get_id at 0x000001C9AE9748B0>
            keyorder   = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': 0, 'f-39aa0784-78a0-447e-9b91-cc3db878996d': 1, 'f-3c4c16a1-469f-4008-8591-034a0c7b6f24': 2, 'f-4286a9dd-d761-4954-8bcb-02a13404601e': 3, ...}
            kwargs     = {}
            loads      = <built-in function loads>
            num_workers = 8
            pack_exception = <function pack_exception at 0x000001C9AE975E10>
            posttask_cbs = ()
            pretask_cbs = ()
            queue      = <queue.Queue object at 0x000001C9AE8B20E0>
            raise_exception = <function reraise at 0x000001C9AE700E50>
            rerun_exceptions_locally = False
            result     = [['f-f1b064dd-9685-45d8-aadb-c851fb5529dd'], ['f-08f5e46d-c78e-4229-a07d-2246bc072fda'], ['f-9c0b1231-9fc3-4bf5-a042-f...-43be-afde-6593999f34cc'], ['f-bc67018d-f89c-45a1-90b6-856c9662a957'], ['f-39aa0784-78a0-447e-9b91-cc3db878996d'], ...]
            result_flat = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda', 'f-39aa0784-78a0-447e-9b91-cc3db878996d', 'f-3c4c16a1-469f-4008-8591-034a0c...-d761-4954-8bcb-02a13404601e', 'f-9c0b1231-9fc3-4bf5-a042-fd04670f73fe', 'f-bc67018d-f89c-45a1-90b6-856c9662a957', ...}
            results    = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda', 'f-39aa0784-78a0-447e-9b91-cc3db878996d', 'f-3c4c16a1-469f-4008-8591-034a0c...-d761-4954-8bcb-02a13404601e', 'f-9c0b1231-9fc3-4bf5-a042-fd04670f73fe', 'f-bc67018d-f89c-45a1-90b6-856c9662a957', ...}
            started_cbs = []
            state      = {'cache': {}, 'dependencies': {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': set(), 'f-39aa0784-78a0-447e-9b91-cc3db878996...733873f-64c6-498c-a396-1bd67a3a7c7a': set(), 'f-f1b064dd-9685-45d8-aadb-c851fb5529dd': set()}), 'finished': set(), ...}
            submit     = <bound method ProcessPoolExecutor.submit of <concurrent.futures.process.ProcessPoolExecutor object at 0x000001C9AE861D80>>
            succeeded  = False
    ..\local.py:491: in fire_tasks
        fut = submit(batch_execute_tasks, each_args)
            _          = 9
            args       = [('f-08f5e46d-c78e-4229-a07d-2246bc072fda', b'\x80\x05\x95}\x02\x00\x00\x00\x00\x00\x00\x8c\x17cloudpickle.cloudpickle...nction loads>, <function _process_get_id at 0x000001C9AE9748B0>, <function pack_exception at 0x000001C9AE975E10>), ...]
            avail_workers = 8
            chunksize  = 6
            data       = {}
            dsk        = {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), 'f...'f-4286a9dd-d761-4954-8bcb-02a13404601e': (<function test_process_initializer.<locals>.f at 0x000001C9AFCCDD80>,), ...}
            dumps      = functools.partial(<function dumps at 0x000001C9AE96EE60>, protocol=5)
            each_args  = [('f-08f5e46d-c78e-4229-a07d-2246bc072fda', b'\x80\x05\x95}\x02\x00\x00\x00\x00\x00\x00\x8c\x17cloudpickle.cloudpickle...in function loads>, <function _process_get_id at 0x000001C9AE9748B0>, <function pack_exception at 0x000001C9AE975E10>)]
            get_id     = <function _process_get_id at 0x000001C9AE9748B0>
            i          = 0
            key        = 'f-f9ed1835-7e08-43be-afde-6593999f34cc'
            loads      = <built-in function loads>
            nready     = 10
            ntasks     = 10
            num_workers = 8
            pack_exception = <function pack_exception at 0x000001C9AE975E10>
            pretask_cbs = ()
            queue      = <queue.Queue object at 0x000001C9AE8B20E0>
            state      = {'cache': {}, 'dependencies': {'f-08f5e46d-c78e-4229-a07d-2246bc072fda': set(), 'f-39aa0784-78a0-447e-9b91-cc3db878996...733873f-64c6-498c-a396-1bd67a3a7c7a': set(), 'f-f1b064dd-9685-45d8-aadb-c851fb5529dd': set()}), 'finished': set(), ...}
            submit     = <bound method ProcessPoolExecutor.submit of <concurrent.futures.process.ProcessPoolExecutor object at 0x000001C9AE861D80>>
            used_workers = 0
    ..\..\..\..\anaconda3\envs\daskenv\lib\concurrent\futures\process.py:705: in submit
        self._adjust_process_count()
            args       = ([('f-08f5e46d-c78e-4229-a07d-2246bc072fda', b'\x80\x05\x95}\x02\x00\x00\x00\x00\x00\x00\x8c\x17cloudpickle.cloudpickl... function loads>, <function _process_get_id at 0x000001C9AE9748B0>, <function pack_exception at 0x000001C9AE975E10>)],)
            f          = <Future at 0x1c9ae8b1db0 state=pending>
            fn         = <function batch_execute_tasks at 0x000001C9AE700AF0>
            kwargs     = {}
            self       = <concurrent.futures.process.ProcessPoolExecutor object at 0x000001C9AE861D80>
            w          = <concurrent.futures.process._WorkItem object at 0x000001C9AE8B1AB0>
    ..\..\..\..\anaconda3\envs\daskenv\lib\concurrent\futures\process.py:683: in _adjust_process_count
        p.start()
            p          = <Process name='Process-3' parent=16580 initial>
            process_count = 0
            self       = <concurrent.futures.process.ProcessPoolExecutor object at 0x000001C9AE861D80>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\process.py:121: in start
        self._popen = self._Popen(self)
            self       = <Process name='Process-3' parent=16580 initial>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\context.py:224: in _Popen
        return _default_context.get_context().Process._Popen(process_obj)
            process_obj = <Process name='Process-3' parent=16580 initial>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\context.py:327: in _Popen
        return Popen(process_obj)
            Popen      = <class 'multiprocessing.popen_spawn_win32.Popen'>
            process_obj = <Process name='Process-3' parent=16580 initial>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\popen_spawn_win32.py:93: in __init__
        reduction.dump(process_obj, to_child)
            cmd        = '"C:\\Users\\runner\\anaconda3\\envs\\daskenv\\python.exe" "-c" "from multiprocessing.spawn import spawn_main; spawn_main(parent_pid=16580, pipe_handle=2224)" "--multiprocessing-fork"'
            env        = None
            hp         = 2364
            ht         = 2352
            pid        = 17452
            prep_data  = {'authkey': b"+'\x99u\x8a\x1c\xe8\x88\xf0\x9de#\x99\xdc\x84h\xdf\x9b\r\x0fQ\xb1\xa9\x1f\xbb[\x89\x97\xaf\xbf8\x8f", 'd...\runner\\.p2\\pool\\plugins\\org.python.pydev.core_9.2.0.202110311311\\pysrc\\runfiles.py', 'log_to_stderr': False, ...}
            process_obj = <Process name='Process-3' parent=16580 initial>
            python_exe = 'C:\\Users\\runner\\anaconda3\\envs\\daskenv\\python.exe'
            rhandle    = 2224
            self       = <multiprocessing.popen_spawn_win32.Popen object at 0x000001C9AE8B1B10>
            tid        = 20184
            to_child   = <_io.BufferedWriter name=5>
            wfd        = 5
            whandle    = 668
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    obj = <Process name='Process-3' parent=16580 initial>
    file = <_io.BufferedWriter name=5>, protocol = None

        def dump(obj, file, protocol=None):
            '''Replacement for pickle.dump() using ForkingPickler.'''
    >       ForkingPickler(file, protocol).dump(obj)
    E       AttributeError: Can't pickle local object 'test_process_initializer.<locals>.proc_init'

    file       = <_io.BufferedWriter name=5>
    obj        = <Process name='Process-3' parent=16580 initial>
    protocol   = None

    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\reduction.py:60: AttributeError
    ============================ slowest 10 durations =============================
    1.87s call     dask/tests/test_multiprocessing.py::test_process_initializer

    (2 durations < 0.005s hidden.  Use -vv to show these durations.)
    =========================== short test summary info ===========================
    FAILED test_multiprocessing.py::test_process_initializer - AttributeError: Ca...
    ============================== 1 failed in 3.11s ==============================

</details>

Reply:
It looks like only functions defined at the top-level can be pickled.  So I tried using the `global` keyword to declare the function first.  But that didn't work either.  (Though the error there was different: "concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending".)
```python
def test_process_initializer():
    global proc_init
    def proc_init():
        global_.value = 1
    
    ... # rest of the code
```
Perhaps this is a limitation of `pickle`?

Reply:
Suspect this comes down to the fact that `pickle` doesn't handle closures whereas `cloudpickle` does. Just using a top-level function instead of a closure (as is done currently) seems like the easy and reasonable fix. So wouldn't worry about this

Reply:
Suspect this comes down to the fact that `pickle` doesn't handle closures whereas `cloudpickle` does. Just using a top-level function instead of a closure (as is done currently) seems like the easy and reasonable fix. So wouldn't worry about this